### PR TITLE
feat: support response hint

### PR DESCRIPTION
**Pull request by dapeng-wang**

### CODE REVIEW

**Comment by kristian** [comment]
RoutingContext routingContext
...
DataContext context

DataRequest request = new DataRequest("target");
request.putHeader("...", "...")

- <DataResponse<T>> requireData(...)

DataResponse
data() --> <T>
headers() --> Map<String, Object>

- DataContext.class:

Map<String, Object> data();
  mergeData(...) // AUTO_MERGE / MANUAL_MERGE, in den DataVerticle eine protected mergeContextData() { ... }

Map<String, Object> responseData(request)

<-- responseData Map<String, Object>, DataContextImpl Map<DataRequest, Map<String, Object>>

Future<String> retriveData(DataQuery, DataMap, DataContext context) {
	...
	DataRequest request = new DataRequest("target", new DataQuery());
	requestData(request, context).compose(result -> {
		Map<String, Object> globalgalaktischeMap = context.data();

		Map<String, Object> responseData = context.responseData(request);
		context.putResponseData("context-type-hint", responseData.get("context-type-hint");
        });
	...
}

**Comment by kristian** [comment]
As promised I made my mind up about the naming scheme here. As we discussed we have a "response" channel of the verticles that want to send data to the upstream verticle (currently handled via `context.responseData().put()) and then for the upstream / downstream verticle to access this information (currently handled via `context.downstreamResponseData().get()`. One of the main problems here is for me confusing both of them, because depending on the "direction of thinking" in your head downstream / upstream, or inbound / outbound or response / requesst, etc. can always be confused. For example:

Does "responseData" mean the data that "was the response" of the verticles I called or is it the "data that I as a verticle being called will respond with"? (at the moment it is the latter).

And this can be applied to all combinations. Stuff like "downstreamResponseData" is kind of redundant and you also again are in the position of the viewer or the sender here, right: So am I as a the consumer of the verticle really "downstream"? Because I would argue, only in the eye of the receiving verticle that so to speak sets the data, I am downstream... So this makes naming for me very difficult...

And based on our todays discussion @dapeng-wang  I think your proposal is the best:

``` 
context.reponseData().put(...) // to return data to the callee, whilst .get() etc. still work like in a normal map and
context.receivedData().get(...) // to read data after the called returned from teh called verticle, whilst .put() on this map should result in an unsupportedoperation exception (so make the map unmodifable) to also clearly indiciate to the end-user that this is only the "read-side" interface of the both
```

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:115]: not needed, everything that is to be changed on the responseData should be changed on the responseData() map itself, e.g. responseData().putAll(...)

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:125]: Same

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:123]: have you considered using a DataMap<Object> here?

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:131]: this should never be possible to be called by an end-user, so please remove

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:155]: if receivedData() would return da DataMap, we could get rid of all these methods as well and the interfaces of DataMap and DataContext would stay consistent regardless of what we do.

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:160]: How about "propagateReceivedData()"?

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:78]: We decided to not staticially import methods that do not describe what they create, e.g.:

Map.of
List.of
Optional.of

Etc. otherwise there could be ambiguities

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:277]: why was all of this moved? this makesreviewing next to impossible, can you move it back and move it in another change if required?

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:313]: As we now have a context.propagateresponseData() method, do we really need this? Seem like overkill to me: So in order to do the propagation, I would have to:

```java
@Override
public ResponseDataPropagationPolicy getResponseDataPropagationPolicy() {
        return ResponseDataPropagationPolicy.AUTO_MERGE;
}
```

And this causes some "magic" to happen. Instead I could simply (in my already implemented `retrieveData` call:

```java
@Override
public Future<?> retrieveData() {
        // ...
        context.propagateResponseData();
        // ...
}
```

Before returning the method. Doesn't the second one fully cover the whole functionallity much more easily and more clear to the end user, because he can influence when and where he wants to do the propagation?

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:395]: Same here no default import of "of" methods please. (We should add that to our coding conventions in the docs folder please)

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:514]: all these lines have been moved around, review is impossible

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:548]: ? unnecesarry

**Review by kristian — CHANGES_REQUESTED** [review]
Ich bin nicht durch gekommen

File: src/main/java/io/neonbee/data/DataContext.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/DataContext.java:123]: I don't think DataMap is a good fit, since we are dealing with key value pairs.

It is unnessary to wrap the key value pairs into an AsyncResult.

File: src/main/java/io/neonbee/data/DataContext.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/DataContext.java:131]: A setter is but required to set the received data in DataVerticle (RecursiveResolutionRoutine) into the DataContext.

Since DataContextImpl is not in the same package as DataVerticle, a package protected method will not work.

File: src/main/java/io/neonbee/data/DataContext.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/DataContext.java:155]: See above.

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/DataVerticle.java:313]: Done

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/DataVerticle.java:277]: The root cause is that some rule was introduced to enforced the order of the public protected and private methods. I didn't check exactly whether in PMD oder Sonar.

Without moving the code, my PR build always failed.

I don't like it either.

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:116]: Same here? Better fit into DataContextImpl, correct? No user should call this, this is only to be called by the framework

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:150]: Maybe just a little text suggestion here:

Propagate all received data intothe current verticle's response data.

@see #responseData

**Review by kristian — CHANGES_REQUESTED** [review]

File: src/main/java/io/neonbee/data/DataContext.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataContext.java:111]: this JavaDoc could be better explaining how responseData / receivedData is to be used and also that some helper methods are available, e.g. findReceivedData, etc.

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:353]: I don't entirely understand this one? as soon as data is received back, we always merge the responseData we received into our own responseData? isn't it that we have to put this into our receivedData instead?! only if propagation is done via the propageteResponseData method then this may be called, isn't it?

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:351]: decodeFromString should not be called twice here, because it is kind of a heavy operation, but maybe it is not needed at all, see comment below

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:563]: does it make sense to only already store the receivedData here, then we could maybe spare us the stream().map operation of the map in line 573-575 or why else is the whole map of the context required?

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (kristian) [inline src/main/java/io/neonbee/data/DataVerticle.java:575]: Could may be be spared out, see comment above

File: src/main/java/io/neonbee/data/internal/DataContextImpl.java
Comment (kristian) [inline src/main/java/io/neonbee/data/internal/DataContextImpl.java:58]: you could put this one above to the other KEY constants also here it still says Metadata, so suggestion would be:

private static final String RESPONSE_DATA_KEY = "responseData";

File: src/main/java/io/neonbee/data/internal/DataContextImpl.java
Comment (kristian) [inline src/main/java/io/neonbee/data/internal/DataContextImpl.java:136]: typo and the resposneData that is passed in might not be mutable, so similar to setData better create a mutable copy?

File: src/main/java/io/neonbee/data/internal/DataContextImpl.java
Comment (kristian) [inline src/main/java/io/neonbee/data/internal/DataContextImpl.java:328]: does this work? as we have a map of Map<String, Object> we can use the new JsonObject(Map) constructor here... I don't think making a put that is a map is supported by JsonObject is it?

also same comment as for adela in the other change, see comment in the `encDetail` line at: https://github.com/SAP/neonbee/pull/165#discussion_r955231128

I think we should make it clear to the user int he JavaDoc comment of responseData(), that the only supported values of the Map<String, Object> are values which are supported by the JsonObject(Map) constructor in cluste roperation mode.

**Review by kristian — CHANGES_REQUESTED** [review]
Very good change DaPeng! Thanks for all the work put into this and sorry for the "concept phase" that held you up :) we try to improve that moving forward!

Just some smaller things and change lgtm after that! 👍 thanks again!

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/DataVerticle.java:563]: I don't think so, we are keeping the reference to the DataContext before sending the requests via the CompositeFuture. At this time, we don't have reference to the receivedData, since they are not yet allocated at this time. The only reference we have is the DataContext wrapper.

File: src/main/java/io/neonbee/data/internal/DataContextImpl.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/internal/DataContextImpl.java:328]: We have been using this code put(DATA_KEY, context.data()) since years, which is also a Map<String, Object>.

Maybe we never exchanged other types than basic types and String.

Nevertheless, I added the map to Jsonobject conversion.

File: src/main/java/io/neonbee/data/DataVerticle.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/DataVerticle.java:353]: the data context will be encoded and decoded when tranferred via event-bus. Here we are receiving the decocded data with updates from eventbus.

File: src/main/java/io/neonbee/data/internal/DataContextImpl.java
Comment (dapeng-wang) [reply src/main/java/io/neonbee/data/internal/DataContextImpl.java:58]: Done

**Comment by s4heid** [comment]
@dapeng-wang please squash these commits together and we can merge :) thank you

**Pull request merged**
